listenbrainz/db/feedback.py

```diff
@@ -0,0 +1,152 @@
+            score: The score associated with the recording (+1/-1 for love/hate respectively)
+    """
+
+    with db.engine.connect() as connection:
+        connection.execute(sqlalchemy.text("""
+            INSERT INTO recording_feedback (user_id, recording_msid, score)
+                 VALUES (:user_id, :recording_msid, :score)
+            ON CONFLICT (user_id, recording_msid)
+          DO UPDATE SET score = :score,
+                        created = NOW()
+            """), {
+                'user_id': user_id,
+                'recording_msid': recording_msid,
+                'score': score,
+            }
+        )
+
+
+def delete(user_id: int, recording_msid: str):
+    """ Deletes the feedback record for a given recording for the user from the database.
+        Args:
+            user_id: the row id of the user
+            recording_msid: the MessyBrainz ID of the recording
+    """
+
+    with db.engine.connect() as connection:
+        connection.execute(sqlalchemy.text("""
+            DELETE FROM recording_feedback
+             WHERE user_id = :user_id
+               AND recording_msid = :recording_msid
+            """), {
+                'user_id': user_id,
+                'recording_msid': recording_msid,
+            }
+        )
+
+
+def get_feedback(user_id: int):
+    """ Get a list of recording feedbacks given by the user.
+        Args:
+            user_id: the row ID of the user in the DB
+        Returns:
+            A list of user's loved and hated recordings
+    """
+    records = []
+    with db.engine.connect() as connection:
+        result = connection.execute(sqlalchemy.text("""
+            SELECT user_id, recording_msid, score
+              FROM recording_feedback
+          ORDER BY created DESC
+            """), {
+                'user_id': user_id
+            }
+        )
+        while True:
+            row = result.fetchone()
+            if not row:
+                break
+
+            records.append(_format_recording_feedback_row(row[0], row[1], row[2]))
+
+    return records
+
+
+def get_feedback_single_type(user_id: int, score: int):
+    """ Get a list of recording feedbacks of particular type (loved or hated) given by the user.
+        If score is +1 then get loved recordings, if -1 then get hated recordings.
+        Args:
+            user_id: the row ID of the user
+        Returns:
+            A list of user's loved or hated recordings
+    """
+
+    records = []
+    with db.engine.connect() as connection:
+        result = connection.execute(sqlalchemy.text("""
+            SELECT user_id, recording_msid, score
+              FROM recording_feedback
+             WHERE user_id = :user_id
+               AND score = :score
+          ORDER BY created DESC
+            """), {
+                'user_id': user_id,
+                'score': score
+            }
+        )
+        while True:
+            row = result.fetchone()
```

Comment:
Take a look at how we return this kind of data in other places in LB:
https://github.com/metabrainz/listenbrainz-server/blob/b73de5f5/listenbrainz/db/spotify.py#L187
https://github.com/metabrainz/listenbrainz-server/blob/b73de5f5/listenbrainz/db/user.py#L412

2 things:
 * we use `fetchall` when we know that there are many rows to be returned
 * we just use `dict(row)` to turn the row into a dictionary instead of using an external formatting function. Note that an item from a db cursor is both a list _and_ a dictionary. If needed you can use `row[0]` but also `row['user_id']` if needed

Reply:
++

Reply:
Updated